Add AKS reference architecture (#2683)
gitpod-io/website: gitpod/docs/self-hosted/latest/reference-architecture/_chunks/networking.md

With Route53 created, you can now proceed to install cert-manager. Cert-manager is needed for Gitpod's internal networking even if you are managing DNS yourself.

</div>

<div slot="azure">

This section will create an Azure managed zone, grant the AKS cluster permission to manage records in that zone and install external-dns.

Create a new Azure managed zone. For example, if you plan on hosting Gitpod at `gitpod.svcs.example.com` then create a managed zone called `svcs.example.com`.

```bash
DOMAIN_NAME="svcs.example.com"
az network dns zone create --name $DOMAIN_NAME --resource-group $RESOURCE_GROUP
```

Authorize the AKS cluster to control DNS records in the zone:

```bash
ZONE_ID=$(az network dns zone show --name "${DOMAIN_NAME}" --resource-group "${RESOURCE_GROUP}" --query "id" -o tsv)
KUBELET_OBJECT_ID=$(az aks show --name "${CLUSTER_NAME}" --resource-group "${RESOURCE_GROUP}" --query "identityProfile.kubeletidentity.objectId" -o tsv)

> KUBELET_CLIENT_ID=$(az aks show --name "${CLUSTER_NAME}" --resource-group "${RESOURCE_GROUP}" --query "identityProfile.kubeletidentity.clientId" -o tsv)

Author: `KUBELET_CLIENT_ID` is used in the context of external-dns and the cert-manager clusterissuer; this needs to be moved.

az role assignment create \
    --assignee "${KUBELET_OBJECT_ID}" \
    --role "DNS Zone Contributor" \
    --scope "${ZONE_ID}"
```

> This role assignment uses [AKS Kubelet Identity](https://cert-manager.io/docs/configuration/acme/dns01/azuredns/#managed-identity-using-aks-kubelet-identity)

> > to authorizes the entire AKS cluster to manage DNS records in the given zone, including cert-manager and external-dns. Other pods (and potentially gitpod workspaces)

Author: Flagging this as a potential concern; will follow-up on this shortly.

Author: Calico's network policy explicitly blocks 169.254.169.254 and Azure Managed Identity relies on the [metadata endpoint](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/tutorial-linux-vm-access-arm) to expose this information. The current approach is secure.

> may be able to alter DNS records as well.

> Install the external-dns helm chart:

Reviewer: ```suggestion
Install the external-dns Helm chart:
```

```

</div>

<div slot="azure">

Install cert-manager with the following command:

```bash
helm repo add jetstack https://charts.jetstack.io
helm repo update
helm upgrade \
    --install \
    --atomic \
    --wait \
    --cleanup-on-fail \
    --create-namespace \
    --namespace='cert-manager' \
    --reset-values \
    --set installCRDs=true \
    --set 'extraArgs={--dns01-recursive-nameservers-only=true,--dns01-recursive-nameservers=8.8.8.8:53\,1.1.1.1:53}' \
    cert-manager \
    jetstack/cert-manager
```

</div>
</CloudPlatformToggle>

### TLS certificate

In this reference architecture, we use cert-manager to also create **TLS certificates for the Gitpod domain**. Since we need wildcard certificates for the subdomains, you must use the [DNS-01 challenge](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge).

> Using a certificate issued by Let's Encrypt is recommended as it minimizes overhead involving TLS certificates and managing CA certificate trust, but is not required. If you already have a TLS certificates for your Gitpod installation with suitable DNS names you can skip this step and use your own certificates during the installation.

Reviewer: ```suggestion
Using a certificate issued by Let's Encrypt is recommended as it minimizes overhead involving TLS certificates and managing CA certificate trust, but is not required. If you already have TLS certificates for your Gitpod installation with suitable DNS names you can skip this step and use your own certificates during the installation.
```

```

</div>

<div slot="azure">

This section will create a cert-manager ClusterIssuer that will generate publicly trusted certificates using Let's Encrypt.

First, determine your Azure subscription ID. You can typically determine your subscription ID from your Azure CLI credentials.

```bash
AZURE_SUBSCRIPTION_ID="$(az account subscription list --query '[0].subscriptionId' --output tsv)"
```

> Then create an `issuer.yaml` containing the following content, expanding the `$AZURE_SUBSCRIPTION_ID`, `$RESOURCE_GROUP`, and `$DOMAIN_NAME` variables:

Author: Poor wording

```suggestion
Then create a file named `issuer.yaml` containing the following content, expanding the `$AZURE_SUBSCRIPTION_ID`, `$RESOURCE_GROUP`, and `$DOMAIN_NAME` variables:
```

```yaml
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: gitpod-issuer
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    email: "email@gitpod.example.com"
    privateKeySecretRef:
      name: issuer-account-key
    solvers:
      - dns01:
          azureDNS:
            subscriptionID: $AZURE_SUBSCRIPTION_ID
            resourceGroupName: $RESOURCE_GROUP
            hostedZoneName: $DOMAIN_NAME
```

Then apply the ClusterIssuer resource:

```bash
kubectl apply -f issuer.yaml
```

> This example ClusterIssuer depends on [Azure Managed Identity](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview) to authorize requests from cert-manager to the AzureDNS API.

> > Refer to the [cert-manager AzureDNS DNS01](https://cert-manager.io/docs/configuration/acme/dns01/azuredns/) documentation for more information on cert-manager API authorization

Author: Missing a trailing period.
